eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php

```diff
@@ -57,492 +57,252 @@ public function createRole(Role $role)
-        )->leftJoin(
-            $this->handler->quoteTable('ezpolicy'),
-            $query->expr->eq(
-                $this->handler->quoteColumn('role_id', 'ezpolicy'),
-                $this->handler->quoteColumn('id', 'ezrole')
-            )
-        )->leftJoin(
-            $this->handler->quoteTable('ezpolicy_limitation'),
-            $query->expr->eq(
-                $this->handler->quoteColumn('policy_id', 'ezpolicy_limitation'),
-                $this->handler->quoteColumn('id', 'ezpolicy')
-            )
-        )->leftJoin(
-            $this->handler->quoteTable('ezpolicy_limitation_value'),
-            $query->expr->eq(
-                $this->handler->quoteColumn('limitation_id', 'ezpolicy_limitation_value'),
-                $this->handler->quoteColumn('id', 'ezpolicy_limitation')
-            )
-        )->where(
-            $query->expr->eq(
-                // Column name "version" is misleading as it stores originalId when creating a draft from an existing role.
-                // But hey, this is legacy! :-)
-                $this->handler->quoteColumn('version', 'ezrole'),
-                $query->bindValue($roleId, null, \PDO::PARAM_STR)
+    public function loadRoleByIdentifier(
+        string $identifier,
+        int $status = Role::STATUS_DEFINED
+    ): array {
+        $query = $this->getLoadRoleQueryBuilder();
+        $query
+            ->where(
+                $query->expr()->eq(
+                    'r.name',
+                    $query->createPositionalParameter($identifier, ParameterType::STRING)
+                )
             )
-        );
+            ->andWhere(
+                $this->buildRoleDraftQueryConstraint($status, $query)
+            );
 
-        $statement = $query->prepare();
-        $statement->execute();
+        $statement = $query->execute();
 
-        return $statement->fetchAll(\PDO::FETCH_ASSOC);
+        return $statement->fetchAll(FetchMode::ASSOCIATIVE);
     }
 
-    /**
-     * Loads all roles.
-     *
-     * @param int $status One of Role::STATUS_DEFINED|Role::STATUS_DRAFT
-     *
-     * @return array
-     */
-    public function loadRoles($status = Role::STATUS_DEFINED)
+    public function loadRoleDraftByRoleId(int $roleId): array
     {
-        $query = $this->handler->createSelectQuery();
-        if ($status === Role::STATUS_DEFINED) {
-            $roleVersionCondition = $query->expr->eq(
-                $this->handler->quoteColumn('version', 'ezrole'),
-                $query->bindValue($status, null, \PDO::PARAM_INT)
-            );
-        } else {
-            $roleVersionCondition = $query->expr->neq(
-                $this->handler->quoteColumn('version', 'ezrole'),
-                $query->bindValue($status, null, \PDO::PARAM_INT)
+        $query = $this->getLoadRoleQueryBuilder();
+        $query
+            ->where(
+            // BC: "version" stores originalId when creating a draft from an existing role
```

Comment:
Nitpick: incorrect indent


```suggestion
                // BC: "version" stores originalId when creating a draft from an existing role
```